community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/storage/paged/InputResources.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.impl.index.storage.paged;
+
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+
+import java.io.IOException;
+import java.util.LinkedList;
+
+import org.neo4j.io.pagecache.PageCursor;
+import org.neo4j.io.pagecache.PagedFile;
+
+/**
+ * Lucene extensively uses {@link IndexInput#clone()} and {@link IndexInput#slice(String, long, long)},
+ * creating clone inputs. Those clones originate from some root input, the input that came from
+ * {@link Directory#openInput(String, IOContext)}.
+ * <p>
+ * It apparently only closes the root inputs, so they become responsible for closing resources associated with the
+ * clones. This resource tracking is off-loaded here.
+ * <p>
+ * The root input has an instance of {@link RootInputResources}, and the clones all share an instance of
+ * {@link CloneInputResources}.
+ * <p>
+ * Each input has it's own {@link PageCursor}, which needs closing, and all inputs from the same root share a
+ * {@link PagedFile}. Cursors are closed if/when the input associated with the cursor is closed, and everything
+ * still open is closed when the root input is closed.
+ * <p>
+ *
+ * @see "ByteBufferIndexInput#close()"
+ */
+public interface InputResources
+{
+    /**
+     * Resource tracking used by Root Inputs - inputs created by {@link Directory#openInput(String, IOContext)}.
+     */
+    class RootInputResources implements InputResources
+    {
+        private final PagedFile pagedFile;
+        private final CloneInputResources cloneResources = new CloneInputResources( this );
+        private LinkedList<PagedIndexInput> clones = new LinkedList<>();
```

Comment:
Haha, I guess because I just needed `append` and `iterate`, so it seemed like an efficient option.. are you thinking b/c of the heap overhead? I guess we do end up creating a hella lot of these index input objects.. 

The MMap implementation from Lucene uses `WeakIdentityMap` for this (see `ByteBufferIndexInput`), so I don't think we're adding overhead vs. current status quo.. 

WDYT of switching to have `PagedIndexInput.next`? Or something different?

Reply:
lets do `PagedIndexInput.next`:)

Reply:
Done! 

Note that I extended `IOUtils` to offer an option for `closeAll` that doesn't allocate arrays; otherwise removing the linked list here kind of is undone by having to convert the `PagedInput.next`-based list into something `closeAll` can consume. Now the `PagedInput.close` is GC-free. 

Might be worth looking at keeping `PagedIndexInput` on a freelist in future diffs, when I was testing for Lucenes `close` behavior I noted it opened several hundred inputs for a simple read query..